### PR TITLE
[GRPH-3] port cli tests from bitshares to Peerplays 

### PR DESCRIPTION
**Pull request by srpatel19590**
Incorporated cli tests from bitshares and related bug fixes and/or enhancements. It requires to add https://github.com/peerplays-network/peerplays/pull/99 to run `chain_test`

### CODE REVIEW

**Comment by oxarbitrage** [comment]
This is the same as https://github.com/peerplays-network/peerplays/pull/98

**Comment by srpatel19590** [comment]
@oxarbitrage GRPH-3 was assigned to me. I've raised this PR as part of it. Moreover, it covers more tests compare to https://github.com/peerplays-network/peerplays/pull/98

**Comment by bobinson** [comment]
@srpatel19590  - Which are the additional tests that you mentioned ?

**Comment by srpatel19590** [comment]
@bobinson Here is the list of additional tests.

- saving_keys_wallet_test
- cli_multisig_transaction
- account_history_pagination
- cli_confidential_tx_test
- cli_set_voting_proxy
- cli_cant_get_signers_from_modified_transaction
- cli_get_available_transaction_signers
- cli_get_signed_transaction_signers

File: libraries/wallet/wallet.cpp [hunk 1]
```diff
@@ -1162,6 +1175,20 @@ class wallet_api_impl
 
       return _builder_transactions[transaction_handle] = sign_transaction(_builder_transactions[transaction_handle], broadcast);
    }
+
+   pair<transaction_id_type,signed_transaction> broadcast_transaction(signed_transaction tx)
```
Comment (bobinson) [inline libraries/wallet/wallet.cpp]: @srpatel19590  - Please provide details of this code block addition.

File: libraries/wallet/include/graphene/wallet/wallet.hpp [hunk 2]
```diff
@@ -590,6 +595,12 @@ class wallet_api
        */
       bool    load_wallet_file(string wallet_filename = "");
 
+      /** Quitting from BitShares wallet.
```
Comment (bobinson) [inline libraries/wallet/include/graphene/wallet/wallet.hpp]: @srpatel19590  - This should be Peerplays wallet

File: libraries/wallet/wallet.cpp [hunk 3]
```diff
@@ -4182,6 +4292,22 @@ signed_transaction wallet_api::sign_transaction(signed_transaction tx, bool broa
    return my->sign_transaction( tx, broadcast);
 } FC_CAPTURE_AND_RETHROW( (tx) ) }
 
+signed_transaction wallet_api::add_transaction_signature( signed_transaction tx,
```
Comment (bobinson) [inline libraries/wallet/wallet.cpp]: @srpatel19590  - Add the reason for addition of this via comment above this commit.

File: tests/CMakeLists.txt [hunk 4]
```diff
@@ -25,6 +25,16 @@ file(GLOB APP_SOURCES "app/*.cpp")
 add_executable( app_test ${APP_SOURCES} )
 target_link_libraries( app_test graphene_app graphene_account_history graphene_bookie graphene_net graphene_chain graphene_egenesis_none fc ${PLATFORM_SPECIFIC_LIBS} )
 
+file(GLOB CLI_SOURCES "cli/*.cpp")
+add_executable( cli_test ${CLI_SOURCES} )
+if(WIN32)
+   list(APPEND PLATFORM_SPECIFIC_LIBS ws2_32)
```
Comment (bobinson) [inline tests/CMakeLists.txt]: @srpatel19590  "PLATFORM_SPECIFIC_LIBS ws2_32" -- what are these libraries ?

File: libraries/chain/get_config.cpp [hunk 5]
```diff
@@ -52,6 +52,7 @@ fc::variant_object get_config()
    result[ "GRAPHENE_DEFAULT_MAINTENANCE_SKIP_SLOTS" ] = GRAPHENE_DEFAULT_MAINTENANCE_SKIP_SLOTS;
    result[ "GRAPHENE_MIN_UNDO_HISTORY" ] = GRAPHENE_MIN_UNDO_HISTORY;
    result[ "GRAPHENE_MAX_UNDO_HISTORY" ] = GRAPHENE_MAX_UNDO_HISTORY;
+   result[ "GRAPHENE_MAX_NESTED_OBJECTS" ] = GRAPHENE_MAX_NESTED_OBJECTS;
```
Comment (bobinson) [inline libraries/chain/get_config.cpp]: @srpatel19590 Why is the nested objects changed ?

File: libraries/chain/include/graphene/chain/config.hpp [hunk 6]
```diff
@@ -53,6 +53,8 @@
 #define GRAPHENE_MIN_UNDO_HISTORY 10
 #define GRAPHENE_MAX_UNDO_HISTORY 10000
 
+#define GRAPHENE_MAX_NESTED_OBJECTS (200)
```
Comment (bobinson) [inline libraries/chain/include/graphene/chain/config.hpp]: @srpatel19590  - provide details of why the nested objects are added.

File: libraries/wallet/wallet.cpp (hunk 1 above)
Comment (srpatel19590) [reply libraries/wallet/wallet.cpp]: It is being called from `cli_multisig_transaction` cli test

```
// case2: broadcast without signature
      // expect: exception with missing active authority
      BOOST_CHECK_THROW(con.wallet_api_ptr->broadcast_transaction(transfer_tx2), fc::exception);
```

File: libraries/wallet/wallet.cpp (hunk 3 above)
Comment (srpatel19590) [reply libraries/wallet/wallet.cpp]: Comment is added in wallet.hpp as below:
Given a fully-formed transaction with or without signatures, signs the transaction with the owned keys and optionally broadcasts the transaction.

This function is called from `cli_multisig_transaction` cli test

File: tests/CMakeLists.txt (hunk 4 above)
Comment (srpatel19590) [reply tests/CMakeLists.txt]: These are windows socket lib. Additionally added to PLATFORM_SPECIFIC_LIBS when the target is windows.

File: libraries/chain/include/graphene/chain/config.hpp (hunk 6 above)
Comment (srpatel19590) [reply libraries/chain/include/graphene/chain/config.hpp]: It is not required. I've already removed this in my next commit.

File: libraries/chain/get_config.cpp (hunk 5 above)
Comment (srpatel19590) [reply libraries/chain/get_config.cpp]: It is not required. I've already removed this in my next commit.

**Comment by srpatel19590** [comment]
@bobinson The two tests `cli_confidential_tx_test` and `account_history_pagination` will be enabled when https://github.com/peerplays-network/peerplays/pull/112 and  https://github.com/peerplays-network/peerplays/pull/109 is merged respectively.

File: tests/CMakeLists.txt (hunk 4 above)
Comment (bobinson) [reply tests/CMakeLists.txt]: @srpatel19590  - provide specific details. ie link, versions etc.

File: libraries/chain/include/graphene/chain/config.hpp (hunk 6 above)
Comment (bobinson) [reply libraries/chain/include/graphene/chain/config.hpp]: Ok

File: tests/CMakeLists.txt (hunk 4 above)
Comment (srpatel19590) [reply tests/CMakeLists.txt]: @bobinson By compiling with Visual Studio 2017 and 2015 Update 1, the required ws2_32 lib will already available and we just required to link it. So MSVS 2017 and 2015 Update 1 is fine.

Comment (srpatel19590) [reply tests/CMakeLists.txt]: We can remove this part if we are not using WIN32. Moreover we can use below instruction to compile in windows.
https://github.com/bitshares/bitshares-core/wiki/BUILD_WIN32#windows-development-environment-setup-and-build-instructions

**Comment by oxarbitrage** [comment]
As we merged https://github.com/peerplays-network/peerplays/pull/98 which similar i think this pull should be closed and open a new one for the test cases we left out of we need them.

**Comment by srpatel19590** [comment]
@bobinson @oxarbitrage PR is updated with just additional cli tests and its fixes

**Comment by bobinson** [comment]
send a clean PR after addressing the dependencies.

**Pull request closed without merging**